DSMR Binding for OH2

mvolaart opened the pull request:
Based on a question of @metbril (https://github.com/openhab/openhab2-addons/issues/1608) i have developed the binding for OH2.

In the past period it has been tested and used by several people (see also https://community.openhab.org/t/new-dsmr-binding-for-openhab-2/20369)

I think it is now ready to merged. Please let me know if i need to address additional items.

martinvw commented:
Can you squash your commits and change commit message to include a `Signed-off-by` (to certify that you wrote it or otherwise have the right to pass it on as an open-source), see http://docs.openhab.org/developers/contributing/contributing.html#sign-your-work

For squashing / rebasing see:
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

martinvw commented:
Please include the binding in the parent pom.xml this way we also have some preliminary checks from the static code analysis.

Maybe also check up on the coding/development guidelines:

http://docs.openhab.org/developers/development/guidelines.html
http://docs.openhab.org/developers/development/bindings.html

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRMeterHandler.java:
```diff
@@ -0,0 +1,205 @@
+            try {
+                meterType = DSMRMeterType.valueOf(getThing().getThingTypeUID().getId().toUpperCase());
+            } catch (Exception exception) {
+                logger.error("Invalid meterType", exception);
```
We could also do the action from line 122 here and return, not much use in continuing. Note that you also log twice

mvolaart replied:
Agreed and fixed

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRMeterHandler.java:
```diff
@@ -0,0 +1,205 @@
+ * The MeterHandler will create logic DSMR meter ThingTypes
+ */
+public class DSMRMeterHandler extends BaseThingHandler implements DSMRMeterListener {
+    // logger
```
Please add only useful comments and if you document fields or methods use proper docblocks.

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRMeterHandler.java:
```diff
@@ -0,0 +1,205 @@
+    private final Logger logger = LoggerFactory.getLogger(DSMRMeterHandler.class);
+
+    // The DSMRMeter instance
+    private DSMRMeter meter = null;
```
Leave out the default `null` value

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRMeterHandler.java:
```diff
@@ -0,0 +1,205 @@
+    private DSMRMeter meter = null;
+
+    // Timestamp when last values were received
+    private long lastValuesReceivedTs = 0;
```
Leave it out as well, you actually did not receive anything at that moment as well ;-).

mvolaart replied:
DOne

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRMeterHandler.java:
```diff
@@ -0,0 +1,205 @@
+                public void run() {
+                    if (System.currentTimeMillis()
+                            - lastValuesReceivedTs > DSMRMeterConstants.METER_VALUES_RECEIVED_TIMEOUT) {
+                        if (!getThing().getStatus().equals(ThingStatus.OFFLINE)) {
```
ThingStatus.OFFLINE is an enum, compare with !=
Why is the check needed what will break by marking an offline device offline?

mvolaart replied:
The reason for this is that i can see from the BaseThingHandler implementation updates to the same state are passed to a callback handler. I don't have a good overview where this handler may come from so i decided to filter unnecessary transitions out.
If this is not necessary (or even harmfull for a good working system) i can take it out.

Please let me know

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRMeterHandler.java:
```diff
@@ -0,0 +1,205 @@
+        if (cosemValues.size() > 0) {
+            lastValuesReceivedTs = System.currentTimeMillis();
+
+            if (!getThing().getStatus().equals(ThingStatus.ONLINE)) {
```
Same questions as line 112

mvolaart replied:
The same reason as before. Filtering out unnecessary status updates

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRMeterHandler.java:
```diff
@@ -0,0 +1,205 @@
+    @Override
+    public void handleRemoval() {
+        // Stop the timeout timer
+        if (meterWatchdog != null) {
```
Better call dispose instead of duplicating
@kaikreuzer is this a common pattern and should it maybe be the responsibility of the framework?

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/DSMRHandlerFactory.java:
```diff
@@ -0,0 +1,115 @@
+                if (thingTypeUIDIsMeter) {
+                    logger.debug("{} is a supported DSMR Meter thing", thingTypeUID);
+                } else {
+                    logger.debug("{} is not a DSMR Meter thing or not a supported DSMR Meter thing", thingTypeUID);
```
If you do not support it please do not log anything maybe on trace but not higher, you will be called for each thing...

mvolaart replied:
I know it will be called for each thing (hence the statement logger.isDebugEnabled to prevent always performing a check to only distuingish in the log statement) but i agree trace is the better solution. I changed this.

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,481 @@
+    private DSMRDeviceConfiguration deviceConfiguration = null;
+
+    // Device status class
+    private class DeviceStatus {
```
Can this be a static inner class?

mvolaart replied:
Made it static (also checked other inner classes and made them static were possible)

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,481 @@
+    private final DeviceStatus deviceStatus;
+
+    // DSMR Port
+    private DSMRPort dsmrPort = null;
```
Please remove the `null` assignments.

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,481 @@
+                        break;
+                    default:
+                        logger.warn("Unknown event {}", event);
+
```
remove newline, remove break

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,481 @@
+                    logger.error("No device state listener available, binding will not work properly!");
+                }
+            } else {
+                logger.debug("Event won't be processed dueo to state {}", deviceStatus.deviceState);
```
Typo 'dueo'

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,481 @@
+        private DeviceState deviceState;
+
+        // Timestamp when the device entered the deviceState
+        private long stateEnteredTS;
```
Please make the field name as clear as the comment is, now I had to scroll back up here to understand the code.

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,481 @@
+                    break;
+                case ONLINE:
+                    if (deviceStatus.receivedCosemObjects.size() > 0) {
+                        List<CosemObject> cosemObjects = new ArrayList<CosemObject>();
```
Diamond operator

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,481 @@
+
+                    break;
+                case ONLINE:
+                    if (deviceStatus.receivedCosemObjects.size() > 0) {
```
Use ! isEmpty instead

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,481 @@
+
+                    break;
+            }
+            if (deviceStatus.receivedCosemObjects.size() > 0) {
```
You are not (yet) interested in the size here, check with isEmpty

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,481 @@
+        if (cosemObjects.size() > 0) {
+            logger.info("There are unhandled CosemObjects, start autodetecting meters");
+
+            DSMRMeterDetector meterDetector = new DSMRMeterDetector();
```
Should / could we reuse the instance of the `DSMRMeterDetector`

mvolaart replied:
I agree creating the instance once is a better idea. Fixed

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,481 @@
+                logger.debug("Event won't be processed dueo to state {}", deviceStatus.deviceState);
+            }
+        }
+    }
```
This some serious big method, it could profit from some sensible splitting up

mvolaart replied:
I split up determing the new state and actually setting the new state

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,481 @@
+            if (discoveryListener != null) {
+                for (DSMRMeterDescriptor meterDescriptor : detectedMeters) {
+                    if (discoveryListener.meterDiscovered(meterDescriptor)) {
+                    } else {
```
Empty if body, why, please add a comment in the code to explain, maybe make it more explicit:

```java
boolean success = discoveryListener.meterDiscovered(meterDescriptor);
if (!success) {
    logger.info("DiscoveryListener {} rejected meter descriptor {}", discoveryListener, meterDescriptor);
}
```

mvolaart replied:
Agree. I took over your suggestion

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,481 @@
+    }
+
+    @Override
+    public void P1TelegramReceived(List<CosemObject> cosemObjects, String telegramDetails) {
```
Please follow naming conventions, should be p1TelegramReceived

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,481 @@
+
+    @Override
+    public void P1TelegramReceived(List<CosemObject> cosemObjects, String telegramDetails) {
+        if (cosemObjects.size() > 0) {
```
use empty

mvolaart replied:
Done. Also searched and replaced all other occurences of .size() > 0 with !isEmpty

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDeviceConfiguration.java:
```diff
@@ -0,0 +1,58 @@
+     * @return true if both are equal, false otherwise or if other == null
+     */
+    @Override
+    public boolean equals(Object other) {
```
Don't implement equals without overriding hashCode

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDeviceConstants.java:
```diff
@@ -0,0 +1,89 @@
+         *
+         * @param eventDetails String containing the details about the event
+         */
+        private DSMRPortEvent(String eventDetails) {
```
private is the default for constructor of enum's you can leave it out.

mvolaart replied:
Done. Removed all private modifier for Enum constructors.

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDeviceConstants.java:
```diff
@@ -0,0 +1,89 @@
+        WRONG_BAUDRATE("Wrong baudrate"),
+        ERROR("General error");
+
+        // Public accessible details about the event
```
Better add a getter for it and make it private.

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDeviceConstants.java:
```diff
@@ -0,0 +1,89 @@
+        SHUTDOWN("DSMR device shutdown");
+
+        // Public accessible details about the event
+        public String eventDetails;
```
Please private final with a getter

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDeviceConstants.java:
```diff
@@ -0,0 +1,89 @@
+         *
+         * @param eventDetails String containing the details about the event
+         */
+        private DSMRDeviceEvent(String eventDetails) {
```
Does not have to be private

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDeviceConstants.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    // Serial port read time out (15 seconds)
+    public static final int SERIAL_PORT_READ_TIMEOUT = (int) TimeUnit.SECONDS.toMillis(15);
```
It's quite some time away but there is a valid reason it is long, can it be long or will you get trouble on places where you use it.

https://en.wikipedia.org/wiki/Year_2038_problem

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRPort.java:
```diff
@@ -0,0 +1,498 @@
+
+    private class DSMRPortStatus {
+        private boolean isOpen = false;
+        private boolean bi = false;
```
please use proper names :-)

mvolaart replied:
I think you refer to bi, oe, fe, pe. In the JavaDoc of SerialPortEvent also these names are used. Even in UART tutorials these abbreviations are used. For this reason i would like to keep these abbreviations. But if you insist :)

martinvw replied:
Getsie ;-) But agree on your reasoning.

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRPort.java:
```diff
@@ -0,0 +1,498 @@
+
+    /* serial port resources */
+    private SerialPort serialPort;
+    private BufferedInputStream bis;
```
Can you use more describing name

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRPort.java:
```diff
@@ -0,0 +1,498 @@
+                dsmrPortListener.handleDSMRPortEvent(DSMRPortEvent.IN_USE);
+            } catch (UnsupportedCommOperationException ucoe) {
+                logger.error(
+                        "Port does not support requested port settings " + "(invalid dsmr:portsettings parameter?): {}",
```
Make this a single string

martinvw replied:
Can you compare your log levels with the guidelines, these should be more like warnings or given that you also change the thing status maybe nothing at all.

See [E.4 @ the coding guidelines](http://docs.openhab.org/developers/development/guidelines.html#e-logging)

mvolaart replied:
I interpretated the coding guidelines describing 'The system' as 'the binding' which is indeed not correct. 

I agree this should be a warning. I would like to keep the log to have the detailed information of the UnsupportedCommOperationException.

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRPort.java:
```diff
@@ -0,0 +1,498 @@
+                }
+                portStatus.pe = seEvent.getNewValue();
+                break;
+            default: /* do nothing */
```
Very big method, maybe the cases should have their own methods with a clear name stating what they do

mvolaart replied:
I had split up this method

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRPort.java:
```diff
@@ -0,0 +1,498 @@
+            }
+        }
+    }
+}
```
Add newline at the end of the file.

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/cosem/CosemDate.java:
```diff
@@ -0,0 +1,128 @@
+        private final Pattern pattern;
+
+        /* Cached java date formatter */
+        private final SimpleDateFormat formatter;
```
Please note that `SimpleDateFormat` is NOT thread safe, Java-8's DateTimeFormatter is your friend in this case.

mvolaart replied:
Thank you for pointing this out. I changed it

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/cosem/CosemHexString.java:
```diff
@@ -0,0 +1,65 @@
+    @Override
+    protected String parse(String cosemHexValue) throws ParseException {
+        if (cosemHexValue.length() % 2 != 0) {
+            throw new ParseException(cosemHexValue + "is not a valid hexadecimal string", 0);
```
Missing space:

> cosemHexValue + "is

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/cosem/CosemObjectFactory.java:
```diff
@@ -0,0 +1,144 @@
+    private final Logger logger = LoggerFactory.getLogger(CosemObjectFactory.class);
+
+    /* internal lookup cache */
+    private final HashMap<OBISIdentifier, CosemObjectType> obisLookupTableFixed;
```
Make the type a `Map`

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/cosem/CosemObjectFactory.java:
```diff
@@ -0,0 +1,144 @@
+         * To facilitate autodiscovery the list has all supported CosemObjectTypes. To improve performance once the
+         * correct OBISIdentifier is discovered for a certain OBISMsgType this is added to the obisLookupTableDynamic.
+         */
+        obisLookupTableFixed = new HashMap<OBISIdentifier, CosemObjectType>();
```
Diamond operator

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/cosem/OBISIdentifier.java:
```diff
@@ -0,0 +1,243 @@
+    public boolean reducedOBISIdentifierIsWildCard() {
+        return (groupA == null) || (groupB == null) || (groupC == null);
+    }
+}
```
Add newline at the end of the file.

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/p1telegram/CRC16.java:
```diff
@@ -0,0 +1,130 @@
+        }
+        return table;
+    }
+}
```
Add newline

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRBridgeDiscoveryHelper.java:
```diff
@@ -0,0 +1,228 @@
+            if (status.state == DiscoveryState.DISCOVERY_RUNNING) {
+                switch (event) {
+                    case DISCOVERY_START:
+                        discoveryTimers.schedule(new Runnable() {
```
Can these be scheduled in the normal scheduler or do you need a lot of them or let them run continuously?

mvolaart replied:
For this timer i use the already available scheduler of AbstractDiscoveryService

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRBridgeDiscoveryHelper.java:
```diff
@@ -0,0 +1,228 @@
+            handleDiscoveryEvent(DiscoveryEvent.BRIDGE_DISCOVERED);
+        }
+    }
+}
```
Missing newline

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRBridgeDiscoveryService.java:
```diff
@@ -0,0 +1,98 @@
+
+        return true;
+    }
+}
```
Missing newline

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRMeterDetector.java:
```diff
@@ -0,0 +1,75 @@
+        }
+        return new LinkedList<DSMRMeterDescriptor>(detectedMeters.values());
+    }
+}
```
Missing newline

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRMeterDiscoveryService.java:
```diff
@@ -0,0 +1,94 @@
+    private final Logger logger = LoggerFactory.getLogger(DSMRMeterDiscoveryService.class);
+
+    // The Bridge ThingUID
+    private ThingUID dsmrBridgeUID = null;
```
Do not set, you can even make it final since it arrives via the constructor

mvolaart replied:
agreed and done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRMeterDiscoveryService.java:
```diff
@@ -0,0 +1,94 @@
+
+        return true;
+    }
+}
```
Missing newline

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/meter/DSMRMeter.java:
```diff
@@ -0,0 +1,151 @@
+        List<CosemObject> filteredList = new LinkedList<>();
+
+        logger.trace("supported identifiers: {}, searching for objects {}", supportedIdentifiers, cosemObjects);
+        for (CosemObject cosemObject : cosemObjects) {
```
You could check out the java 8 stream().filter() for this

mvolaart replied:
I took over your suggestion. Thanks

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/meter/DSMRMeter.java:
```diff
@@ -0,0 +1,151 @@
+     * @return true if the identification and metertype are equal
+     */
+    @Override
+    public boolean equals(Object other) {
```
With equals you should always implement hashcode, please check all occurrences of either equals and hashcode.

mvolaart replied:
Done. Checked also all other occurences

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/meter/DSMRMeterConfiguration.java:
```diff
@@ -0,0 +1,44 @@
+     * @return true if both are equal, false otherwise or if other == null
+     */
+    @Override
+    public boolean equals(Object other) {
```
equals

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/meter/DSMRMeterConstants.java:
```diff
@@ -0,0 +1,26 @@
+ */
+public class DSMRMeterConstants {
+    // Timeout for receiving meter values (in seconds)
+    public static final int METER_VALUES_RECEIVED_TIMEOUT = (int) TimeUnit.SECONDS.toMillis(60);
```
Any reason it cannot be a long

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/meter/DSMRMeterDescriptor.java:
```diff
@@ -0,0 +1,77 @@
+     * @return true if both objects are equal, false otherwise
+     */
+    @Override
+    public boolean equals(Object other) {
```
equals

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/meter/DSMRMeterKind.java:
```diff
@@ -0,0 +1,47 @@
+    SLAVE_ELECTRICITY2("Slave electricity meter 2");
+
+    // name of the meterkind
+    private String name;
```
Make final

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/meter/DSMRMeterKind.java:
```diff
@@ -0,0 +1,47 @@
+    SLAVE_ELECTRICITY1("Slave electricity meter"),
+    SLAVE_ELECTRICITY2("Slave electricity meter 2");
+
+    // name of the meterkind
```
Please check all comments like this and turn into real docblocks

mvolaart replied:
Done

martinvw commented on addons/binding/org.openhab.binding.dsmr/pom.xml:
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
To have the binding being picked up by the distro, you furthermore need to add it to the feature.xml, again at the alphabetically correct position. If you have a dependency on some transport bundle (e.g. upnp, mdns or serial), make sure to add a line for this dependency as well (see the other bindings as an example).

mvolaart replied:
Done

mvolaart commented on addons/binding/org.openhab.binding.dsmr/ESH-INF/config/configuration_parameters.xml:
```diff
@@ -0,0 +1,36 @@
+    xmlns:config-description="http://eclipse.org/smarthome/schemas/config-description/v1.0.0"
+    xsi:schemaLocation="http://eclipse.org/smarthome/schemas/config-description/v1.0.0 http://eclipse.org/smarthome/schemas/config-description-1.0.0.xsd">
+
+  <config-description uri="binding:dsmr:bridgesettings">
```
i reformatted all the other xml files too

mvolaart commented on addons/binding/org.openhab.binding.dsmr/ESH-INF/config/configuration_parameters.xml:
```diff
@@ -0,0 +1,36 @@
+        <description>The DSMR-device channel for this meter. This is a technical value filled by the binding. Users should not adapt this.</description>
+    </parameter>
+  </config-description>
+</config-description:config-descriptions>
```
All xml files checked and added newlines where necessary

mvolaart commented on addons/binding/org.openhab.binding.dsmr/ESH-INF/thing/dsmr_device.xml:
```diff
@@ -0,0 +1,13 @@
+        <description>The Dutch Smart Meter device</description>
+        <config-description-ref uri="binding:dsmr:bridgesettings"/>
+    </bridge-type>
+    
```
fixed

mvolaart commented on addons/binding/org.openhab.binding.dsmr/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,26 @@
+Bundle-Name: DSMR Binding
+Bundle-SymbolicName: org.openhab.binding.dsmr;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.0.0.qualifier
```
Done

mvolaart commented on addons/binding/org.openhab.binding.dsmr/README.md:
```diff
@@ -0,0 +1,294 @@
+# DSMR Binding
```
I didn't know this. Fixed

mvolaart commented on addons/binding/org.openhab.binding.dsmr/README.md:
```diff
@@ -0,0 +1,294 @@
+This binding reads the P1-port of the Dutch Smart Meters that comply to NTA8130, DSMR v2.1, DSMR v2.2, DSMR v3.0, DSMR v4.0, DSMR v4.04, and DSMR 5.0.  
+Although DSMR v4.2 is not an official specification, the binding has support for this version.
+
+If you are not living in the Netherlands but do want to read a meter please have look at the [IEC-62056-21 Meter Binding](https://github.com/openhab/openhab1-addons/wiki/IEC-62056---21-Meter-Binding).
```
I'm not the maintainer of this binding, so i don't know if it is ported already, or if it is WIP.

The link is correct (i.e. it refers to valid page) and i think we need to have something here since as far as i know only people in the Netherlands do have equipment that will be working with this binding. If there is already an OH2 implementation of this generic meter binding please let me know. I will update the URL.

martinvw replied:
Good that you checked, I think this it is the most recent version

mvolaart commented on addons/binding/org.openhab.binding.dsmr/README.md:
```diff
@@ -0,0 +1,294 @@
+#### Configuration
+The configuration consists of the following parameters:
+- `serialPort` (mandatory). The serial port where the P1-port is connected to (e.g. Linux: `/dev/ttyUSB1`, Windows: `COM2`)
+- `serialPortSettings` (optional, default: &lt;empty&gt;). Serial Port parameters in the format `<speed> <nr of bits><parity [E(ven)/N(one)/O(dd)]><stop bits (1/1.5/2)>`.  
```
I agree. I will make an update for this

mvolaart commented on addons/binding/org.openhab.binding.dsmr/README.md:
```diff
@@ -0,0 +1,294 @@
+
+Supported meters:  
+
+| Meter Thing | Thing type ID | M-Bus channel | Refresh rate |
```
Done

mvolaart commented on addons/binding/org.openhab.binding.dsmr/README.md:
```diff
@@ -0,0 +1,294 @@
+- O channel is supported only if the device has this functionality
+
+
+| Channel Type ID | Item Type | Description | Ace4000 | DSMR V2.1 | DSMR V2.2 | DSMR V3.0 | DSMR V4.0 | DSMR V4.0.4 | DSMR V4.2 | DSMR V5 |
```
Done

mvolaart commented on addons/binding/org.openhab.binding.dsmr/about.html:
```diff
@@ -0,0 +1,28 @@
+and such source code may be obtained at <a href="http://www.openhab.org/">openhab.org</a>.</p>
+
+</body>
+</html>
```
Done

mvolaart commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/DSMRBindingConstants.java:
```diff
@@ -0,0 +1,35 @@
+    // The other Thing Types are defined in DSMRMeterTypes
+
+    // Discovery timeout
+    public final static int DSMR_DISCOVERY_TIMEOUT = 60;
```
Done

mvolaart commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRBridgeHandler.java:
```diff
@@ -0,0 +1,220 @@
+ * @since 2.1.0
+ */
+public class DSMRBridgeHandler extends BaseBridgeHandler implements DSMRDeviceStateListener {
+    // logger
```
I agree that is does not add any additional information :) I removed those comments

mvolaart commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRBridgeHandler.java:
```diff
@@ -0,0 +1,220 @@
+    private final Logger logger = LoggerFactory.getLogger(DSMRBridgeHandler.class);
+
+    // DSMRDevice that belongs to this DSMRBridgeHandler
+    private DSMRDevice dsmrDevice = null;
```
Fixed. Also removed other occurences

mvolaart commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRBridgeHandler.java:
```diff
@@ -0,0 +1,220 @@
+    private DSMRMeterDiscoveryListener discoveryService;
+
+    // The available meters for this bridge
+    private final List<DSMRMeter> availableMeters = new ArrayList<DSMRMeter>();
```
I'm aware of the diamond operator :) I updated all the occurences

mvolaart commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRMeterHandler.java:
```diff
@@ -0,0 +1,205 @@
+        DSMRMeterConfiguration meterConfig = null;
+        DSMRMeterDescriptor meterDescriptor = null;
+
+        if (config != null) {
```
No i did not ran into troubels. I agree about the documentation and removed the unecessary check

mvolaart commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRMeterHandler.java:
```diff
@@ -0,0 +1,205 @@
+        DSMRMeterDescriptor meterDescriptor = null;
+
+        if (config != null) {
+            meterConfig = config.as(DSMRMeterConfiguration.class);
```
Agreed and removed the check

mvolaart commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRMeterHandler.java:
```diff
@@ -0,0 +1,205 @@
+
+            try {
+                meterType = DSMRMeterType.valueOf(getThing().getThingTypeUID().getId().toUpperCase());
+            } catch (Exception exception) {
```
Agreed and fixed

mvolaart commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRBridgeHandler.java:
```diff
@@ -0,0 +1,220 @@
+    // logger
+    private final Logger logger = LoggerFactory.getLogger(DSMRBridgeHandler.class);
+
+    // DSMRDevice that belongs to this DSMRBridgeHandler
```
Done

mvolaart commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRBridgeHandler.java:
```diff
@@ -0,0 +1,220 @@
+
+    @Override
+    public void handleRemoval() {
+        if (watchdog != null && !watchdog.isCancelled()) {
```
Done

mvolaart commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRBridgeHandler.java:
```diff
@@ -0,0 +1,220 @@
+        DSMRDeviceConfiguration deviceConfig = config.as(DSMRDeviceConfiguration.class);
+
+        logger.debug("Using configuration {}", deviceConfig);
+        if (deviceConfig == null || deviceConfig.serialPort == null || deviceConfig.serialPort.length() == 0) {
```
I took over your suggestion. Thanks.

mvolaart commented on addons/binding/org.openhab.binding.dsmr/build.properties:
```diff
@@ -0,0 +1,6 @@
+bin.includes = META-INF/,\
+               .,\
+               OSGI-INF/,\
+               ESH-INF/
```
Done

mvolaart commented on addons/binding/org.openhab.binding.dsmr/ESH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,48 @@
+        xsi:schemaLocation="http://eclipse.org/smarthome/schemas/binding/v1.0.0 http://eclipse.org/smarthome/schemas/binding-1.0.0.xsd">
+
+    <name>DSMR Binding</name>
+    <description><![CDATA[ <h3>General information</h3><p>
```
I think it looks pretty readable. I have removed some parts, since there were not relevant.

mvolaart commented on addons/binding/org.openhab.binding.dsmr/ESH-INF/config/configuration_parameters.xml:
```diff
@@ -0,0 +1,36 @@
+    <parameter name="serialPortSettings" type="text" pattern="\d+\s+\d[neoNEO](1\.5|1|2)" required="false">
+      <default></default>
+      <label>Serial Port Settings</label>
+      <description><![CDATA[ In case autodetecting the Serial Port settings fails it is possible to specify the Serial Port Settings in this field.<br>
```
Yes they do

mvolaart commented:
@martinvw I have wrapped up the open points where i have some questions

[Q] How does this look in the paperui interface, normally this just contains a very short description.
[A] I removed some text. It is still some information, but readable. Please let me know if it must be really short

[Q] DSMRBridgeHandler - stateChanged:  is this the responsibility of each and every binding... (question to Kaikreuzer)
[A] Is there is already an answer yet? Do i have to update this?

[Q] DSMRMeterHandler - initialize & meterValueReceived: Why is the check needed what will break by marking an offline device offline?
[A] The reason for this is that i can see from the BaseThingHandler implementation updates to the same state are passed to a callback handler. I don't have a good overview where this handler may come from so i decided to filter unnecessary transitions out.
If this is not necessary (or even harmfull for a good working system) i can take it out.

Please let me know

[Q] DSMRPort class DSMRPortStatus: please use proper names :-)
[A] I think you refer to bi, oe, fe, pe. In the JavaDoc of SerialPortEvent also these names are used. Even in UART tutorials these abbreviations are used. For this reason i would like to keep these abbreviations. But if you insist... ;-)

[Q] DSMRBridgeDiscoveryHelper: Can these be scheduled in the normal scheduler or do you need a lot of them or let them run continuously?
--> I don't quite understand your question. I use ThreadPoolManager to create two thread pools. One to start asynchronous tasks and the other to start timer tasks
The only reason why i came to this implementation was D.2 of the coding guidelines (Creation of threads must be avoided. Instead, resort into using existing schedulers which use pre-configured thread pools).

So my questions:
What do you mean with 'normal scheduler'? 
Is there a generic thread pool name that must be used?
Did i totally misinterpret the goal of D2?

Some help needed :-)

martinvw commented:
> >How does this look in the paperui interface, normally this just contains a very short description.

> I removed some text. It is still some information, but readable. Please let me know if it must be really short

<img width="392" alt="screen shot 2017-06-08 at 17 30 28" src="https://user-images.githubusercontent.com/2625279/26936921-4a91f812-4c70-11e7-82dc-97ac9adc6471.png">

Compare your binding with the others in the paperUI on my system.

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRBridgeHandler.java:
```diff
@@ -0,0 +1,220 @@
+    @Override
+    public void stateChanged(DeviceState oldState, DeviceState newState, String stateDetail) {
+        /*
+         * Only if the Thing is initialized the State may be updated (otherwise an IllegalStateException is thrown)
```
> @kaikreuzer is this the responsibility of each and every binding...

@kaikreuzer can you check this question ^^^

martinvw commented:
> What do you mean with 'normal scheduler'?
Is there a generic thread pool name that must be used?
Did i totally misinterpret the goal of D2?

There are some existing thread pools which should be used if you just schedule something now and then, if you continously needs threads or a single thread you can have your own thread (pool). 

I think the ThingHandler has some protected reference to a thread pool.

martinvw commented:
> [Q] DSMRMeterHandler - initialize & meterValueReceived: Why is the check needed what will break by marking an offline device offline?

I have to read myself back in :-)

martinvw commented:
Did you already push, the comments do not seem to be marked as outdated?

mvolaart commented:
I did not push anything yet (i would like to first address every point and do some testing :-) )

mvolaart commented:
But i will push some commits. I necessary i can squash them after

mvolaart commented:
For some reason i can not reword my first commit from Eclipse to add a "signed-off-by" or squash them with the other commits.

martinvw commented:
@mvolaart thanks!

@kaikreuzer will do a final review as soon as he worked his was through the other bindings in his queue.

metbril commented:
Any progress on this binding? Will it be included in the next release?

@mvolaart where can I download the latest snapshot? Is it already on Cloudbees?

gedejong commented:
Bump, any progress on this binding? I've been using the dev version since the beginning of this year without problems. It is a big improvement over 1.x version. 

@mvolaart thanks so much for all the effort!

martinvw commented:
@mvolaart can you fix the merge conflict, ideally by rebasing.

@kaikreuzer i have it also running for a long time already ;-)

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRMeterHandler.java:
```diff
@@ -0,0 +1,206 @@
+        }, DSMRMeterConstants.METER_VALUES_RECEIVED_TIMEOUT, DSMRMeterConstants.METER_VALUES_RECEIVED_TIMEOUT,
+                TimeUnit.MILLISECONDS);
+
+        updateStatus(ThingStatus.OFFLINE);
```
Why OFFLINE? You should only set it to offline, if it is for sure unreachable. Otherwise set it to UNKNOWN.

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRMeterHandler.java:
```diff
@@ -0,0 +1,206 @@
+     * Remove the Meter Thing
+     */
+    @Override
+    public void handleRemoval() {
```
remove this method

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/DSMRHandlerFactory.java:
```diff
@@ -0,0 +1,116 @@
+ * The {@link DSMRHandlerFactory} is responsible for creating things and thing
+ * handlers.
+ *
+ * @author M. Volaart
```
add "- Initial contribution"
... and same for all further classes, where it is missing.

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/DSMRHandlerFactory.java:
```diff
@@ -0,0 +1,116 @@
+        if (thingTypeUID.equals(DSMRBindingConstants.THING_TYPE_DSMR_BRIDGE)) {
+            Bridge dsmrBridge = (Bridge) thing;
+            DSMRMeterDiscoveryService discoveryService = new DSMRMeterDiscoveryService(dsmrBridge.getUID());
+            DSMRBridgeHandler bridgeHandler = new DSMRBridgeHandler((Bridge) thing, discoveryService);
```
This is actually the wrong way around - a handler must not have a dependency on the discovery service - but a discovery service can use an existing handler. Check the HueThingHandlerFactory as an example: It passes the handler instance to the newly created discovery service:
```
HueLightDiscoveryService discoveryService = new HueLightDiscoveryService(bridgeHandler);
```

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/DSMRHandlerFactory.java:
```diff
@@ -0,0 +1,116 @@
+    /**
+     * The registration handler used for notifying new detected meters
+     */
+    private ServiceRegistration<?> serviceReg;
```
A single one isn't enough - you are creating a service for every single bridge and there might be more than one. You'll need to keep a Map here, check `HueThingHandlerFactory` as an example.

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,507 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * The DSMRDevice class represents the physical Thing within OpenHAB2 context
```
remove "within OpenHAB2 context" (or alternatively correct the spelling to "openHAB" ;-))

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,507 @@
+ * If there are communication problems or no P1 telegrams are received (anymore), the device will enter the state
+ * OFFLINE.
+ *
+ * If the OpenHAB2 system wants the binding to shutdown, the DSMRDevice enters the SHUTDOWN state and will close
```
remove OpenHAB2

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,507 @@
+ *
+ * In case of configuration errors the DSMRDevice will enter the state CONFIGURATION_PROBLEM
+ *
+ * @author M. Volaart
```
add "- Initial contribution"

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,507 @@
+             * No update of state of received telegrams for a period of RECOVERY_TIMEOUT
+             * Evaluate current state
+             */
+            executor.submit(new Runnable() {
```
change to
```
executor.submit(() -> { handleDeviceState(); });
```

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/DSMRBindingConstants.java:
```diff
@@ -0,0 +1,42 @@
+    /**
+     * Constant for the DSMR Thread pool
+     */
+    public static final String DSMR_THREAD_POOL_NAME = "DSMR Thread Pool";
```
Normally, bindings must not create their own thread pool, but reuse the existing "scheduler".
Do we have a special situation here which requires us to have a separate pool?

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,507 @@
+                        logger.warn("Unknown event {}", event);
+                }
+                // Handle the current state asynchronous
+                executor.submit(new Runnable() {
```
change to
```
executor.submit(() -> { handleDeviceState(); });
```

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,507 @@
+                        cosemObjects.addAll(deviceStatus.receivedCosemObjects);
+
+                        // Handle cosem objects asynchronous
+                        executor.submit(new Runnable() {
```
change to Java8 syntax

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,507 @@
+                        deviceStatus.receivedCosemObjects.clear();
+                    } else if (System.currentTimeMillis()
+                            - deviceStatus.lastTelegramReceivedTS > DSMRDeviceConstants.RECOVERY_TIMEOUT) {
+                        logger.info("No messages received for at least {} ms, reinitialize device",
```
reduce to debug

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,507 @@
+         * There are still unhandled cosemObjects. Start discovery of DSMR meters
+         */
+        if (!cosemObjects.isEmpty()) {
+            logger.info("There are unhandled CosemObjects, start autodetecting meters");
```
reduce to debug

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,507 @@
+            logger.info("There are unhandled CosemObjects, start autodetecting meters");
+
+            List<DSMRMeterDescriptor> detectedMeters = meterDetector.detectMeters(cosemObjects);
+            logger.info("Detected the following new meters: {}", detectedMeters.toString());
```
reduce to debug

kaikreuzer replied:
remove toString()

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,507 @@
+                for (DSMRMeterDescriptor meterDescriptor : detectedMeters) {
+                    boolean success = discoveryListener.meterDiscovered(meterDescriptor);
+                    if (!success) {
+                        logger.info("DiscoveryListener {} rejected meter descriptor {}", discoveryListener,
```
reduce to debug

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,507 @@
+                    }
+                }
+            } else {
+                logger.warn("There is no listener for new meters!");
```
Is this a software bug that needs to be reported? Otherwise reduce to debug.

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,507 @@
+            deviceStatus.lastTelegramReceivedTS = System.currentTimeMillis();
+            deviceStatus.receivedCosemObjects.addAll(cosemObjects);
+        } else {
+            logger.debug("Parsing was succesful, however there were no CosemObjects");
```
typo: successful

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRPort.java:
```diff
@@ -0,0 +1,576 @@
+
+                        dsmrPortListener.handleDSMRPortEvent(DSMRPortEvent.ERROR);
+                    }
+                    logger.info("DSMR Port opened successful");
```
might be worth to mention on which port you have successfully connected (there might be multiple).

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRPort.java:
```diff
@@ -0,0 +1,576 @@
+                    try {
+                        serialPort.addEventListener(this);
+                    } catch (TooManyListenersException tmle) {
+                        logger.error("DSMR binding will not be operational.", tmle);
```
shouldn't you better return in that case instead of continuing?

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRPort.java:
```diff
@@ -0,0 +1,576 @@
+     */
+    public void close() {
+        synchronized (portLock) {
+            logger.info("Closing DSMR port");
```
reduce to debug

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRPort.java:
```diff
@@ -0,0 +1,576 @@
+            logger.debug("Switched port settings to: {}", portSettings);
+        } else {
+            portSettings = fixedPortSettings;
+            logger.info("Fixed port settings configured (autodetect DISABLED): {}", portSettings);
```
mention the port

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRBridgeDiscoveryHelper.java:
```diff
@@ -0,0 +1,251 @@
+            if (status.state == DiscoveryState.DISCOVERY_RUNNING) {
+                switch (event) {
+                    case DISCOVERY_START:
+                        timerService.schedule(new Runnable() {
```
change to Java8

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRBridgeDiscoveryHelper.java:
```diff
@@ -0,0 +1,251 @@
+                                handleDiscoveryEvent(DiscoveryEvent.DISCOVERY_SWITCHBAUDRATE);
+                            }
+                        }, DSMRDeviceConstants.SERIAL_PORT_AUTO_DETECT_TIMEOUT, TimeUnit.MILLISECONDS);
+                        timerService.schedule(new Runnable() {
```
change to Java8

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRBridgeDiscoveryService.java:
```diff
@@ -0,0 +1,97 @@
+    /**
+     * Starts a new discovery scan.
+     *
+     * All available Serial Ports are scanned for P1 telegrams.
```
Is this really a safe thing to do? Couldn't that have the weirdest impacts on some connected serial devices which are no meters?

mvolaart replied:
I think it can be done because there won't be sent any date to the particular serial port (only reading). If the ports to other devices are already in use we skip these ports. 
So we will try only reading on serial ports that are not claimed by any other software application.

kaikreuzer commented on addons/binding/pom.xml:
```diff
@@ -28,6 +28,7 @@
     <module>org.openhab.binding.coolmasternet</module>
     <module>org.openhab.binding.dlinksmarthome</module>
     <module>org.openhab.binding.dscalarm</module>
+	<module>org.openhab.binding.dsmr</module>
```
please replace tab by spaces

martinvw replied:
@kaikreuzer our code guidelines suggest otherwise (tabs in xml) should we reformat this file once?

kaikreuzer replied:
For files that are not in the IDE workspace I actually do not care too much (they are not considered by our auto-formatting). They should just be in sync with the rest of the file, which does not seem to be the case here.

kaikreuzer commented on features/openhab-addons/src/main/feature/feature.xml:
```diff
@@ -79,6 +79,12 @@
         <bundle start-level="80">mvn:org.openhab.binding/org.openhab.binding.dscalarm/${project.version}</bundle>
     </feature>
+	
+	<feature name="openhab-binding-dsmr" description="DSMR Binding" version="${project.version}">
```
please replace tab by spaces

martinvw commented:
@mvolaart I just noted this info log passing every second:

```

2017-09-26 15:00:18.880 [INFO ] [ding.dsmr.internal.device.DSMRDevice] - There are unhandled CosemObjects, start autodetecting meters
2017-09-26 15:00:18.881 [INFO ] [ding.dsmr.internal.device.DSMRDevice] - Detected the following new meters: [Meter type: M3_V5_0, channel: 1]
2017-09-26 15:00:19.882 [INFO ] [ding.dsmr.internal.device.DSMRDevice] - There are unhandled CosemObjects, start autodetecting meters
2017-09-26 15:00:19.882 [INFO ] [ding.dsmr.internal.device.DSMRDevice] - Detected the following new meters: [Meter type: M3_V5_0, channel: 1]
2017-09-26 15:00:20.914 [INFO ] [ding.dsmr.internal.device.DSMRDevice] - There are unhandled CosemObjects, start autodetecting meters
2017-09-26 15:00:20.914 [INFO ] [ding.dsmr.internal.device.DSMRDevice] - Detected the following new meters: [Meter type: M3_V5_0, channel: 1]
2017-09-26 15:00:21.882 [INFO ] [ding.dsmr.internal.device.DSMRDevice] - There are unhandled CosemObjects, start autodetecting meters
2017-09-26 15:00:21.889 [INFO ] [ding.dsmr.internal.device.DSMRDevice] - Detected the following new meters: [Meter type: M3_V5_0, channel: 1]
2017-09-26 15:00:22.915 [INFO ] [ding.dsmr.internal.device.DSMRDevice] - There are unhandled CosemObjects, start autodetecting meters
2017-09-26 15:00:22.915 [INFO ] [ding.dsmr.internal.device.DSMRDevice] - Detected the following new meters: [Meter type: M3_V5_0, channel: 1]
2017-09-26 15:00:23.884 [INFO ] [ding.dsmr.internal.device.DSMRDevice] - There are unhandled CosemObjects, start autodetecting meters
2017-09-26 15:00:23.885 [INFO ] [ding.dsmr.internal.device.DSMRDevice] - Detected the following new meters: [Meter type: M3_V5_0, channel: 1]
2017-09-26 15:00:24.887 [INFO ] [ding.dsmr.internal.device.DSMRDevice] - There are unhandled CosemObjects, start autodetecting meters
2017-09-26 15:00:24.888 [INFO ] [ding.dsmr.internal.device.DSMRDevice] - Detected the following new meters: [Meter type: M3_V5_0, channel: 1]
```

wdyt?

martinvw commented:
@mvolaart are you still there? It would be great to get this merged 👍 

Do you need help or other support?

Also it seems your binding disappeared from the marketplace, did you do that yourself or did you receive any notification about it?

Hilbrand commented:
@mvolaart I was also wondering what your plans are? I wanted to use this binding, but found it didn't work with the latest snapshot so I cloned and updated it in a branch. I also made some other changes that might be useful, but before changing to much could you give an update on what you plans are?

@martinvw these logs are caused because when data is received and it has known data but no thing that handles it, it will start a discovery process (probably not a desired pattern I guess) I presume in general it will find at minimal a specific device and a generic device. For example a specific energy device contains the meter readings and that is the one you added/configured. The generic device you probably didn't configure, as long as it's not configured it keeps discovering it (for example see discovered devices in paper ui and if you add the generic device the logs will disappear).

mvolaart commented:
Yes i'm still around. Last months i have been very busy.

My plan is for sure to get this binding merged. I will make some time the coming weeks to get this further.

I did not remove anything from the marketplace. I'm not aware of any notification that this has happened (but may be i missed it)

Hilbrand commented:
Great to see you're still working on it. Looking at the binding it doesn't look like the raw serial data isn't parsed differently for different meters. It's just the same data with different content and there is no different code logic for different meters. All differences are in data, right? So you could also have 1 thing with all possible channels, but that would have a lot of channels not in use by a specific device. But would make the code simpler. Now the thing is, in recent changes in smart home it is now possible to dynamically create channels. So instead of what is here, 1 binding and multiple things,  one could have 1   thing that creates channels as found in the data. This would make all much simpler. Maybe something to consider?

martinvw commented:
> It's just the same data with different content and there is no different code logic for different meters. All differences are in data, right? So you could also have 1 thing with all possible channels, but that would have a lot of channels not in use by a specific device. But would make the code simpler.

Maybe a good moment to validate with @kaikreuzer because imho an electrical meter is not the same as a gas meter and should not threaten as a single thing. If that is now possible in smart home it should be used with great caution imho.

kaikreuzer commented:
I agree with @martinvw here that dynamic channels should not be used unless there isn't another good option.
I very much liked it in this PR to see all the different "real" physical meters to be represented as its own thing type, listing exactly the channels they have.
The code should not become more complex through this - all meters share the same channel types and as such the handler really only need to look at the channel type and implement the logic for it once.
If support for a new meter is added, all there is to do is to add a new XML for it - there should be no need to touch any code (maybe with the exception of some constants). See e.g. the addition of a [new Sonos device](https://github.com/eclipse/smarthome/pull/4429/files) recently.

Dynamic channels can be used when the remote device is completely unknown and can potentially dynamically change its feature set. In such a case it should be queried and constructed dynamically.

Hilbrand commented:
Ok, maybe dynamic channels is not such a good idea here ;-) But the reality here is complexer I believe. The things in this binding are mostly not "real" devices but protocol versions with some have a list of optional channels a meter might or might not support. What triggered me was when I used this binding I get 2 things while I only have 1 real meter.

mvolaart commented:
I think some background information is needed so it is more clear why the binding is designed as it is now.

OpenhAB2 Things:"Things represent the physical layer of an openHAB system. From a configuration standpoint, Things tell openHAB which physical entities (devices, web services, information sources, etc.) are to be managed by the system.
The physical entities here are Dutch Smart Meters. There is only 1 meta data field to identify the physical type and that is the DSMR-specification. The DSMR-specification described besides protocol information, also serial port communication settings and update frequency settings. Imho these are different physical devices that also needs there DSMR-identification in the thing. The user can look up this information and i can image he wants to know if the binding has correctly identified his device.
When we look at possible setups of a Dutch Smart Meter the most simple setup is only the meter itself. However the meter supports attaching other meters as well (water, gas, heating, slave electricity, etc.). These other devices are imho also different physical devices and represent also different things. The Smart Meter is in this situation only a 'bridge' to the real meter.
In the more recent specifications of DSMR it becomes more clear the that main electricity meter can be seen as just a regular meter (like water, gas etc.) only bundled in the same case as the 'bridge' (that does the serial communication). 
From this point of view the 'bridge' part of the Smart Meter is more like a Z-Wave USB Stick where the meters are the 'real' devices.

At this moment i don't know yet the concept of dynamic channels exactly but regardless of this i still like the idea that OpenHAB2 identifies the Thing with the DSMR-specification number which the user can find on the Internet when he looks up the documentation of his Smart Meter.

Hilbrand commented:
@mvolaart any update on your plans to move this binding forward? I'm still interested in this binding, but in its current state it doesn't work with the latest version of openHAB. I can help if that's ok with you? Also is it an idea to move the generic thing into the bridge so no artificial thing is needed for the generic channel?

mvolaart commented:
I finally get some time to bring the binding further. I will start looking into the current problems

Hilbrand commented:
Nice to see you're updating! I'll be also giving some feedback. However. I noticed a number of comments from you on Jun 8 confirming you fixed things, but these changes are not to be found in the latest code. Did you by accident push forced over these changes? I'm missing commit: `- Updates from review on pull request #2289` it doesn't seem be squashed either. I still have it in my branch.

mvolaart commented:
I indeed screwed up my set up.

So if you indeed can help me with the link to this commit i will be very thankful.

mvolaart commented on addons/binding/org.openhab.binding.dsmr/ESH-INF/thing/meter_cooling_ace4000.xml:
```diff
@@ -0,0 +1,19 @@
+	xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+
+	<thing-type id="cooling_ace4000" listed="false">
```
There are many variations of the meter devices (hence the list of meter xml files). For a regular user without knowing the DSMR-specifications it is not possible to select the correct device.

Therefor the binding relies fully on autodetection and prevent users adding devices that doesn't or only partial work correctly.

Hilbrand commented:
The commit is in my branch: https://github.com/Hilbrand/openhab2-addons/tree/dsmr_update (It also contains some improvements in additional commits).

mvolaart commented:
@martinvw @kaikreuzer 

Excuse me for the late update of your review comments.

I believe i have now addressed all of the review comments.

Please let me know if someting more has to be done to get the DSMR-binding merged.

Hilbrand commented on addons/binding/org.openhab.binding.dsmr/ESH-INF/thing/channeltypes_cooling.xml:
```diff
@@ -0,0 +1,13 @@
+        xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
+        xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+
+    <channel-type id="coolingValueType">
```
This file also contains spaces instead of tabs.

Hilbrand commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/p1telegram/P1TelegramParser.java:
```diff
@@ -0,0 +1,452 @@
+                             */
+                            FutureTask<Void> task = new FutureTask<>(
+                                    new P1TelegramCallable(cosemObjects, telegramState));
+                            task.run();
```
Not sure what is happening here. But this is a synchronize call. It will only be asynchroon if it's run by an executor service that starts this task in a new thread.

mvolaart replied:
Thank you for pointing this out. The executor service is missing indeed.

Hilbrand commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/meter/DSMRMeter.java:
```diff
@@ -0,0 +1,160 @@
+ * appropriate one.
+ * - {@link DSMRMeterIdentification} describing the identification of this meter.
+ *
+ * {@link DSMRMeterType} and {@link DSMRMeterIdentification} are private members of the DSMRMeter
```
Is this javadoc comment still up-to-date. It suggest there are private members in this class that are not here.

Hilbrand commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/meter/DSMRMeter.java:
```diff
@@ -0,0 +1,160 @@
+        List<CosemObject> filteredList;
+
+        logger.trace("supported identifiers: {}, searching for objects {}", supportedIdentifiers, cosemObjects);
+        filteredList = cosemObjects.stream()
```
This can be simplified to simply return here. The variable `filteredList` isn't really used.

Hilbrand commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/p1telegram/P1TelegramParser.java:
```diff
@@ -0,0 +1,461 @@
+                             */
+                            FutureTask<Void> task = new FutureTask<>(
+                                    new P1TelegramCallable(cosemObjects, telegramState));
+                            p1TelegramService.execute(task);
```
The callable doesn't have any use here. You can simply write this as:
`p1TelegramService.execute(() -> telegramListener.telegramReceived(cosemObjects, telegramState));` and remove the inner class `P1TelegramCallable`.

Also for simplicity is it really necessary to send values asynchrone or can you not just return a value when all data is handled and let the caller handle the values/thrown an exception on error?

mvolaart replied:
Thank you for pointing this out. I already process the values asynchronous in DSMRDevice. Another asynchronous implementation is bit too much

Hilbrand commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,508 @@
+                        newDeviceState = DeviceState.CONFIGURATION_PROBLEM;
+                        break;
+                    case ERROR:
+                        // General error occured
```
`occured` -> `occurred`.

wborn commented:
@mvolaart When I make a local build of this PR and run it my dsmrBridge (P1 Version 42) is continuously going online and offline. See logging below where I grepped on 'dsmrBridge':

```
2018-01-07 13:50:22.061 [hingStatusInfoChangedEvent] - 'dsmr:dsmrBridge:22e5393c' changed from OFFLINE (COMMUNICATION_ERROR): No telegrams received for too long to ONLINE: P1 telegram received OK
2018-01-07 13:51:01.993 [hingStatusInfoChangedEvent] - 'dsmr:dsmrBridge:22e5393c' changed from ONLINE: P1 telegram received OK to OFFLINE (COMMUNICATION_ERROR): No telegrams received for too long
2018-01-07 13:51:11.974 [hingStatusInfoChangedEvent] - 'dsmr:dsmrBridge:22e5393c' changed from OFFLINE (COMMUNICATION_ERROR): No telegrams received for too long to ONLINE: P1 telegram received OK
2018-01-07 13:53:41.864 [hingStatusInfoChangedEvent] - 'dsmr:dsmrBridge:22e5393c' changed from ONLINE: P1 telegram received OK to OFFLINE (COMMUNICATION_ERROR): No telegrams received for too long
2018-01-07 13:53:51.857 [hingStatusInfoChangedEvent] - 'dsmr:dsmrBridge:22e5393c' changed from OFFLINE (COMMUNICATION_ERROR): No telegrams received for too long to ONLINE: P1 telegram received OK
2018-01-07 13:57:31.681 [hingStatusInfoChangedEvent] - 'dsmr:dsmrBridge:22e5393c' changed from ONLINE: P1 telegram received OK to OFFLINE (COMMUNICATION_ERROR): No telegrams received for too long
2018-01-07 13:57:41.701 [hingStatusInfoChangedEvent] - 'dsmr:dsmrBridge:22e5393c' changed from OFFLINE (COMMUNICATION_ERROR): No telegrams received for too long to ONLINE: P1 telegram received OK
2018-01-07 14:00:51.650 [hingStatusInfoChangedEvent] - 'dsmr:dsmrBridge:22e5393c' changed from ONLINE: P1 telegram received OK to OFFLINE (COMMUNICATION_ERROR): No telegrams received for too long
2018-01-07 14:01:01.622 [hingStatusInfoChangedEvent] - 'dsmr:dsmrBridge:22e5393c' changed from OFFLINE (COMMUNICATION_ERROR): No telegrams received for too long to ONLINE: P1 telegram received OK
2018-01-07 14:01:01.639 [hingStatusInfoChangedEvent] - 'dsmr:dsmrBridge:22e5393c' changed from ONLINE: P1 telegram received OK to OFFLINE (COMMUNICATION_ERROR): No telegrams received for too long
2018-01-07 14:01:11.620 [hingStatusInfoChangedEvent] - 'dsmr:dsmrBridge:22e5393c' changed from OFFLINE (COMMUNICATION_ERROR): No telegrams received for too long to ONLINE: P1 telegram received OK
2018-01-07 14:01:11.635 [hingStatusInfoChangedEvent] - 'dsmr:dsmrBridge:22e5393c' changed from ONLINE: P1 telegram received OK to OFFLINE (COMMUNICATION_ERROR): No telegrams received for too long
2018-01-07 14:01:21.586 [hingStatusInfoChangedEvent] - 'dsmr:dsmrBridge:22e5393c' changed from OFFLINE (COMMUNICATION_ERROR): No telegrams received for too long to ONLINE: P1 telegram received OK
2018-01-07 14:01:21.626 [hingStatusInfoChangedEvent] - 'dsmr:dsmrBridge:22e5393c' changed from ONLINE: P1 telegram received OK to OFFLINE (COMMUNICATION_ERROR): No telegrams received for too long
2018-01-07 14:01:31.576 [hingStatusInfoChangedEvent] - 'dsmr:dsmrBridge:22e5393c' changed from OFFLINE (COMMUNICATION_ERROR): No telegrams received for too long to ONLINE: P1 telegram received OK
2018-01-07 14:01:31.581 [hingStatusInfoChangedEvent] - 'dsmr:dsmrBridge:22e5393c' changed from ONLINE: P1 telegram received OK to OFFLINE (COMMUNICATION_ERROR): No telegrams received for too long
2018-01-07 14:01:41.581 [hingStatusInfoChangedEvent] - 'dsmr:dsmrBridge:22e5393c' changed from OFFLINE (COMMUNICATION_ERROR): No telegrams received for too long to ONLINE: P1 telegram received OK
2018-01-07 14:01:41.615 [hingStatusInfoChangedEvent] - 'dsmr:dsmrBridge:22e5393c' changed from ONLINE: P1 telegram received OK to OFFLINE (COMMUNICATION_ERROR): No telegrams received for too long
2018-01-07 14:01:51.593 [hingStatusInfoChangedEvent] - 'dsmr:dsmrBridge:22e5393c' changed from OFFLINE (COMMUNICATION_ERROR): No telegrams received for too long to ONLINE: P1 telegram received OK
2018-01-07 14:01:51.601 [hingStatusInfoChangedEvent] - 'dsmr:dsmrBridge:22e5393c' changed from ONLINE: P1 telegram received OK to OFFLINE (COMMUNICATION_ERROR): No telegrams received for too long
2018-01-07 14:02:01.575 [hingStatusInfoChangedEvent] - 'dsmr:dsmrBridge:22e5393c' changed from OFFLINE (COMMUNICATION_ERROR): No telegrams received for too long to ONLINE: P1 telegram received OK

```

mvolaart commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/p1telegram/P1TelegramParser.java:
```diff
@@ -0,0 +1,461 @@
+    /**
+     * Abort the current telegram
+     */
+    public void abortTelegram() {
```
I agree on the thread safety. This needs definitely some improvements

mvolaart commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRMeterHandler.java:
```diff
@@ -0,0 +1,181 @@
+
+        // Update the internal states
+        if (!cosemValues.isEmpty()) {
+            lastValuesReceivedTs = System.currentTimeMillis();
```
IMHO not. An DSMR meter should always have values available. If the meter has no values available there is probably something wrong. So in this case i think it is ok to not update the value to the watchdog can inform the user no valid values are received.

mvolaart commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,508 @@
+        deviceStatus.deviceStateEnteredTS = System.currentTimeMillis();
+
+        // Handle the new state asynchronous
+        executor.submit(() -> {
```
This is indeed not necessary

mvolaart commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java:
```diff
@@ -0,0 +1,508 @@
+             * No update of state of received telegrams for a period of RECOVERY_TIMEOUT
+             * Evaluate current state
+             */
+            executor.submit(() -> {
```
I agree. It is not necessary

mvolaart commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/meter/DSMRMeterKind.java:
```diff
@@ -0,0 +1,48 @@
+ * @author M. Volaart - Initial contribution
+ */
+public enum DSMRMeterKind {
+    INVALID("Invalid meter"),
```
That's correct. I added this so it can be used when necessary

mvolaart commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/meter/DSMRMeterKind.java:
```diff
@@ -0,0 +1,48 @@
+    HEATING("Heating meter"),
+    COOLING("Cooling meter"),
+    WATER("Water meter"),
+    GENERIC("Generic meter"),
```
A generic meter is a meter defined by DSMR v3. It is a meter that does readings without specifying if this is gas / liquid (with a valve in m3 unit) or  electricity (with a switch in kWh unit).

So this meter will just send values. Hence the name Generic.

mvolaart commented:
@wborn Can you please provide some debug logging?

wborn commented:
It looks like the 30sec communications timeout occurs just when it has finished reinitializing @mvolaart . 
See this [online-offline.log](https://github.com/openhab/openhab2-addons/files/1620578/online-offline.log).

wborn commented on addons/binding/org.openhab.binding.dsmr/OSGI-INF/DSMRDiscoveryService.xml:
```diff
@@ -0,0 +1,20 @@
+    http://www.eclipse.org/legal/epl-v10.html
+
+-->
+<scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.1.0" immediate="true" name="discovery.dsmr">
```
This file should be removed because the DSMRBridgeDiscoveryService class is annotated with `@Component`.

wborn commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/DSMRHandlerFactory.java:
```diff
@@ -0,0 +1,131 @@
+ *
+ * @author M. Volaart - Initial contribution
+ */
+@Component(service = ThingHandlerFactory.class)
```
Please also add the following to the `@Component` annotation: `configurationPid = "binding.dsmr"`

This way users can [override](https://www.eclipse.org/smarthome/documentation/features/dsl.html#default-configuration-file) configuration parameters using the common naming scheme.

wborn commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRBridgeDiscoveryService.java:
```diff
@@ -0,0 +1,99 @@
+ *
+ * @author M. Volaart - Initial contribution
+ */
+@Component(service = DiscoveryService.class)
```
Please also add the following to the `@Component` annotation: `configurationPid = "discovery.dsmr"`

This way users can [override](https://www.eclipse.org/smarthome/documentation/features/dsl.html#default-configuration-file) configuration parameters using the common naming scheme.

kaikreuzer commented:
@mvolaart Do you see any chance to get the review comments addressed by the end of this week? If so, the binding would make it into the 2.3 release, which would be fantastic!

Hilbrand commented:
@mvolaart can you give us some feedback? You've invested a lot of effort in the past years, but maybe don't have the time to finish it at this moment? To let you know I've been making changes to your binding code in my own fork. Because I wanted it to work for me. My code seems fairly stable (but needs some cleaning up, mostly fix javadoc). Among other things I've fixed my review comments. So that work might not have to be done twice. On the other hand, I did make some significant changes in parts of the code that might not necessarily be in line with your ideas of how you want it to work. Let us know what your plans are or if you think my code could be helpful?

Hilbrand commented:
Thanks for the quick response @mvolaart. I'll continue cleanup up my code and hope to have a pr ready in coming weeks. It would be nice to have it in at least the 2.4.0 release. When I'm finished I'll create a new pr.

Regarding your list:
1. I've refactored the discovery already very heavily such that discovery is only done when a discovery is triggered by the user and not during reading of values. This will not have the problem of reporting of found devices during normal operation. There are some more changes in that part of the code that improve stability.
2. I was thinking about having serial-over-ip. HomeAssistant has it. I haven't looked into how it would work, but I've already moved some code around to make it less entangled with the serial specific part, so it should be fairly easy to add such support.

kaikreuzer commented:
> I was thinking about having serial-over-ip

Please see https://github.com/eclipse/smarthome/pull/5560 - the goal is to provide serial-over-IP directly within the framework and now having every individual binding deal with that.

Hilbrand commented:
@kaikreuzer I guess it also would require this binding to use `org.eclipse.smarthome.io.transport.serial`? Although that might be a good idea/needed anyway even if the serial-over-ip is not added.

kaikreuzer commented:
@Hilbrand Yes, this is right - but that should not mean much effort as it is effectively only a different namespace.

martinvw commented:
Closed because of the follow up
